Include/XTaskQueue.h

```diff
@@ -0,0 +1,309 @@
+/// Registers a wait handle with the task queue.  When the wait handle
+/// is satisfied the task queue will invoke the given callback. This
+/// provides an efficient way to add items to a task queue in 
+/// response to handles becomming signaled.
+/// </summary>
+/// <param name='queue'>The queue to submit the callback to.</param>
+/// <param name='port'>The port to invoke the callback on.</param>
+/// <param name='waitHandle'>The handle to monitor.</param>
+/// <param name='callbackContext'>An optional context pointer that will be passed to the callback.</param>
+/// <param name='callback'>A pointer to the callback function.</param>
+/// <param name='token'>A registration token.</param>
+STDAPI XTaskQueueRegisterWaiter(
+    _In_ XTaskQueueHandle queue,
+    _In_ XTaskQueuePort port,
+    _In_ HANDLE waitHandle,
+    _In_opt_ void* callbackContext,
+    _In_ XTaskQueueCallback* callback,
+    _Out_ XTaskQueueRegistrationToken* token
+    ) noexcept;
+
+/// <summary>
+/// Unregisters a previously registered task queue waiter.
+/// </summary>
+/// <param name='queue'>The queue to remove the waiter from.</param>
+/// <param name='token'>The token returned from XTaskQueueRegisterWaiter.</param>
+STDAPI_(void) XTaskQueueUnregisterWaiter(
+    _In_ XTaskQueueHandle queue,
+    _In_ XTaskQueueRegistrationToken token
+    ) noexcept;
+
+/// <summary>
+/// Registers a callback that will be invoked whenever a callback
+/// is submitted to this queue.
+/// </summary>
+/// <param name='queue'>The queue to register the submit callback to.</param>
+/// <param name='callbackContext'>An optional context pointer to be passed to the callback.</param>
+/// <param name='callback'>A callback that will be invoked when a new callback is submitted to the queue.</param>
+/// <param name='token'>A token used in a later call to XTaskQueueUnregisterMonitor to remove the callback.</param>
+STDAPI XTaskQueueRegisterMonitor(
+    _In_ XTaskQueueHandle queue,
+    _In_opt_ void* callbackContext,
+    _In_ XTaskQueueMonitorCallback* callback,
+    _Out_ XTaskQueueRegistrationToken* token
+    ) noexcept;
+
+/// <summary>
+/// Unregisters a previously registered callback. This blocks if there are outstanding montior
+/// callbacks.
+/// </summary>
+/// <param name='queue'>The queue to remove the submit callback from.</param>
+/// <param name='token'>The token returned from XTaskQueueRegisterMonitor.</param>
+STDAPI_(void) XTaskQueueUnregisterMonitor(
+    _In_ XTaskQueueHandle queue,
+    _In_ XTaskQueueRegistrationToken token
+    ) noexcept;
+
+/// <summary>
+/// Returns a handle to the process task queue, or nullptr if there is no
+/// process task queue.  By default, there is a process task queue
+/// that uses the thread pool for both work and completion ports. You
+/// can replace the default process task queue by calling 
+/// XTaskQueueSetCurrentProcessTaskQueue.  There is no need to close the
+/// handle returned from XTaskQueueGetCurrentProcessTaskQueue.
```

Comment:
mmm delicious race conditions

Reply:
Could be. The idea is a process configures this at startup before it begins making calls.  I can see your point though -- if more and more stuff starts using this then the possibility of a race condition before the process queue can be configured is real.